File: Python/new_cnpj.py
import re
from typing import List


class NewCNPJ:
    """
    Validador e formatador para o Novo CNPJ com letras e dígitos.
    """

    @staticmethod
    def is_valid(cnpj: str) -> bool:
        """
        Verifica se o Novo CNPJ é válido.
        Espera 14 caracteres alfanuméricos após limpeza.
        """
        if not isinstance(cnpj, str) or not re.fullmatch(r"[0-9A-Z./-]+", cnpj):
            return False

        cleaned = re.sub(r"[^0-9A-Z]", "", cnpj)
        if len(cleaned) != 14:
            return False

        base = cleaned[:12]
        expected = cleaned[12:]
        calculated = NewCNPJ._calculate_check_digits(base)

        return expected == calculated

    @staticmethod
    def format(cnpj: str) -> str:
        """
        Formata o Novo CNPJ no padrão: XX.XXX.XXX/XXXX-XX
        Retorna string vazia se inválido.
        """
        if not NewCNPJ.is_valid(cnpj):
            return ""

        cleaned = re.sub(r"[^0-9A-Z]", "", cnpj)
        return (
            f"{cleaned[:2]}.{cleaned[2:5]}.{cleaned[5:8]}/"
            f"{cleaned[8:12]}-{cleaned[12:]}"
        )

    @staticmethod
    def _calculate_check_digits(base: str) -> str:
        """
        Calcula os dois dígitos verificadores com pesos personalizados.
        """
        weights1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
        weights2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]

        values = [NewCNPJ._char_to_value(c) for c in base]
        d1 = NewCNPJ._calculate_check_digit(values, weights1)
        d2 = NewCNPJ._calculate_check_digit(values + [d1], weights2)

        return f"{d1}{d2}"

    @staticmethod
    def _calculate_check_digit(values: List[int], weights: List[int]) -> int:
        total = sum(v * w for v, w in zip(values, weights))
        remainder = total % 11
        return 0 if remainder < 2 else 11 - remainder

    @staticmethod
    def _char_to_value(char: str) -> int:
        if "0" <= char <= "9":
            return ord(char) - ord("0")
        elif "A" <= char <= "Z":
            return ord(char) - ord("A") + 17
        return 0

File: Swift/NewCNPJ.swift
import Foundation

class NewCNPJ {
    
    static func isValid(_ cnpj: String) -> Bool {
        let pattern = "^[0-9A-Z./-]+$"
        guard cnpj.range(of: pattern, options: .regularExpression) != nil else {
            return false
        }
        
        let cleanedCNPJ = cnpj.replacingOccurrences(of: "[^0-9A-Z]", with: "", options: .regularExpression)
        guard cleanedCNPJ.count == 14 else { return false }
        
        let base = String(cleanedCNPJ.prefix(12))
        let expectedCheckDigits = String(cleanedCNPJ.suffix(2))
        let calculatedCheckDigits = calculateCheckDigits(base: base)
        
        return expectedCheckDigits == calculatedCheckDigits
    }

    static func format(_ cnpj: String) -> String {
        let cleaned = cnpj.replacingOccurrences(of: "[^0-9A-Z]", with: "", options: .regularExpression)
        guard cleaned.count == 14 else { return "" }
        
        return "\(cleaned.prefix(2)).\(cleaned[2..<5]).\(cleaned[5..<8])/\(cleaned[8..<12])-\(cleaned.suffix(2))"
    }

    private static func calculateCheckDigits(base: String) -> String {
        let weights1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
        let weights2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
        
        let convertedBase = base.map { charToValue($0) }
        let firstDigit = calculateCheckDigit(values: convertedBase, weights: weights1)
        let secondDigit = calculateCheckDigit(values: convertedBase + [firstDigit], weights: weights2)
        
        return "\(firstDigit)\(secondDigit)"
    }

    private static func calculateCheckDigit(values: [Int], weights: [Int]) -> Int {
        let sum = zip(values, weights).map(*).reduce(0, +)
        let remainder = sum % 11
        return remainder < 2 ? 0 : 11 - remainder
    }

    private static func charToValue(_ char: Character) -> Int {
        if char >= "0" && char <= "9" {
            return Int(String(char))!
        } else if char >= "A" && char <= "Z" {
            return Int(char.asciiValue!) - 65 + 17
        } else {
            return 0
        }
    }
}


